core/server/master/src/main/java/alluxio/master/file/loadmanager/LoadManager.java

```diff
@@ -87,393 +83,276 @@ public final class LoadManager {
-                }
-              }
-            }
-          });
+  /**
+   * Get active workers.
+   * @return active workers
+   */
+  @VisibleForTesting
+  public Map<WorkerInfo, CloseableResource<BlockWorkerClient>> getActiveWorkers() {
+    return mActiveWorkers;
+  }
 
-          /*If no workers can be acquired, fail the entire load task immediately.*/
-          if (addresses.isEmpty()) {
-            load.addUfsLoadStatus(load.getPath(), TaskStatus.FAILURE);
-          }
+  /**
+   * Refresh active workers.
+   */
+  @VisibleForTesting
+  public void updateWorkers() {
+    Set<WorkerInfo> workerInfos;
+    try {
+      workerInfos = ImmutableSet.copyOf(mFileSystemMaster.getWorkerInfoList());
+    } catch (UnavailableException e) {
+      LOG.warn("Failed to get worker info, using existing worker infos of {} workers",
+          mActiveWorkers.size());
+      return;
+    }
+    if (workerInfos.size() == mActiveWorkers.size()
+        && workerInfos.containsAll(mActiveWorkers.keySet())) {
+      return;
+    }
 
-          mCurrentSize.decrementAndGet();
-        } catch (InterruptedException e) {
-          Thread.currentThread().interrupt();
+    Map<WorkerInfo, CloseableResource<BlockWorkerClient>> activeWorkers = new HashMap<>();
+    for (WorkerInfo workerInfo : workerInfos) {
+      if (mActiveWorkers.containsKey(workerInfo)) {
+        activeWorkers.put(workerInfo, mActiveWorkers.get(workerInfo));
+      } else {
+        try {
+          activeWorkers.put(
+              workerInfo, mContext.acquireBlockWorkerClient(workerInfo.getAddress()));
+        } catch (IOException e) {
+          // skip the worker if we cannot obtain a client
         }
       }
     }
+    mActiveWorkers = activeWorkers;
+    mWorkerLastUpdateTime = System.currentTimeMillis();
+  }
 
-    void loadBlockBatch(BlockWorkerClient client,
-                                WorkerNetAddress address,
-                                BlockBuffer blockBuffer,
-                                Load load) {
-      if (!blockBuffer.hasNext()) {
-        load.setWorkerLoadFinished(address, true);
-        return;
-      }
-
-      List<Block> blockBatch = blockBuffer.getNextFileBlockBatch();
+  /**
+   * Get load jobs.
+   * @return load jobs
+   */
+  @VisibleForTesting
+  public Map<String, Load> getLoadJobs() {
+    return mLoadJobs;
+  }
 
-      LoadRequest loadRequest = buildRequest(load, blockBatch);
-      ListenableFuture<LoadResponse> listenableFuture = client.load(loadRequest);
-      Futures.addCallback(listenableFuture, new FutureCallback<LoadResponse>() {
-        @Override
-        public void onSuccess(@Nonnull LoadResponse r) {
-          TaskStatus s = r.getStatus();
+  private void run() {
+    while (!Thread.currentThread().isInterrupted()) {
+      if (System.currentTimeMillis() - WORKER_UPDATE_INTERVAL > mWorkerLastUpdateTime) {
+        updateWorkers();
+      }
+      mLoadTasks.forEach(this::processJob);
```

Comment:
this might consume a lot of CPU time even if there's no job. Blocking queue might be a better option here for the producer-consumer pattern.